bots/honkbot.py

```diff
@@ -136,6 +129,24 @@ async def leave(self, ctx, *role: str):
                 "I do not have permissions to assign roles right now. Sorry!"
             )
 
+    def get_role_from_input(self, ctx, role_input: tuple):
+        """
+        Turn a user's string from !join or !leave into a Discord Role object
+
+        :param ctx: Discord Context (for accessing the Roles list)
+        :param role_input: Tuple of Strings representing user input (split by spaces)
+        :return: A tuple of either a Role or False, and an accompanying message
+        """
+        if len(role_input) != 1:
+            return False, "Usage: !join [" + ", ".join(self.custom_roles) + "]"
+
+        role = next((role for role in self.custom_roles if role_input[0].lower() == role.lower()), None)
```

Comment:
I wish this could be cleaner but this is the best I could come up with.

It builds a generator that stops returning the second it finds a role that matches, and returns the correctly capitalized role for that match. If it doesn't find one, it returns None.

Reply:
I think this works for the time being. I think eventually I want to move role lists to the config file rather than having it be hard coded to make the code a little less janky. Then role objects can be generated dynamically from that and added to a tracker object. If we were to move forward with that, then this would have to be a little more robust to deal with potential duplicates, but for now I think it's fine.

Reply:
If the configuration has the locked roles (admin, mod, vip) then the bot itself can get its own allowed roles, and then you can add to it by just adding them in discord. Presumably the locked roles change less often.

Regardless, I think the docstrings have to have the list hardcoded - not sure how to generate those :(